### ISSUE TITLE
classloader error when updating local lib while keeping sbt running

### ISSUE THREAD

**Issue by mpollmeier** [sbt/sbt#3584]
## steps to reproduce
1. `git clone https://github.com/mpollmeier/sbt-publishlocal-bug.git`
1. open two terminals, one in library, one in service
1. in library: `sbt publishLocal`
1. in service: start sbt, run `compile` (everything compiles fine), keep sbt open
1. in library: change anything, e.g. `echo "object A" >> src/main/scala/Library.scala`
1. in library: `sbt publishLocal`
1. in service: run `compile` again

## Result
Fails to load the jar, with different error messages, depending on god knows why. These ones I've seen so far:

```
[error] error while loading Library, invalid stored block lengths

[error] error while loading Library, ZIP_Read: error reading zip file

[error] error while loading Library, class file '/home/mp/.ivy2/local/sbttest/library_2.12/0.1-SNAPSHOT/jars/library_2.12.jar(Library.class)' is broken
[error] (class java.lang.ArrayIndexOutOfBoundsException/627)

error while loading Library, read error
[error] /home/mp/tmp/sbt-publishlocal-bug/service/src/main/scala/Service.scala:3: Library does not have a constructor
```

followed by:
```
[error] /home/mp/tmp/sbt-publishlocal-bug/service/src/main/scala/Service.scala:3:11: Library does not have a constructor
[error]   println(new Library().foo)
[error]           ^
[error] two errors found
```

This problem exists in sbt 0.13.16 and 1.0.2, with both publishLocal and publishM2. 
I searched around and the only related issue I found was https://github.com/sbt/sbt/issues/892 , which is about sbt plugins, and suggests that this is a jvm classloader issue, and that `in general, sbt can cache/reuse class loaders, but it only currently does this for Scala jars, the loaders are behind SoftReferences, and it checks the last modified time before reusing the class loader.`

Since this is such a common use case, I am surprised that there's not more people talking about it - or am I doing something wrong?
If not, it would really be great if we could provide a workaround, e.g. by catching this error and reloading the classloader. I'm happy to help with the implementation if you point me in the right direction.

**Comment by drxcc** [comment]
I have been running up against this issue for several months now.  I was assuming I was doing something wrong when I didn't see anyone else with the issue and it seemed like such a common use case.  

In the past I would essentially continually run `~publishLocal` from one project and `~run` in another and I could edit any file without worrying which project it was from.   I have been using this mode of development for several years and it only stopped working recently; so it has worked in the past.  However, there were several big jumps that I moves that happened at the same time and I thought it could be related to any of the following changes:

1. moved from scala-2.11 to scala-2.12
2. moved from sbt-0.13.14  to sbt-0.13.16 (and 1.0.2)
3. moved from jdk7 to jdk8 

If I have time; I will try to roll back to a set of versions that I did work.

**Comment by mpollmeier** [comment]
Your first idea was right: the problem doesn't exist with scala 2.11.11, which suggests that this isn't actually an sbt problem...

**Comment by drxcc** [comment]
Interesting you have the same issue.  I’m very surprised no one else has run into this.   I was able to verify that my environment does work with jdk8, sbt-0.13.6 and scala-2.11.11 so it seems like a 2.12 related issue.  

Additionally, to reproduce the 2.12 errors, I had to make changes in the upstream project file that depended on classes that have changed in the downstream .  In other words a 2.12 push worked if the there were no dependent changes.  However,  a “clean; publishLocal” always causes a break.

**Comment by mpollmeier** [comment]
I just bisected the officially released scala versions and found that 2.12.0-M3 is the last working version, and 2.12.0-M4 is the first version with this bug. Next step: trace it down to the actual commit.

**Comment by mpollmeier** [comment]
Unfortunately back then the sbt build of the scala repo wasn't as good as it is today. I can run `sbt publishLocal` on the (e.g. revision `d0ca6ed277`) and it publishes as `2.12.0-SNAPSHOT`, but my sample projects don't find it. For later scala versions the same process works just fine :(

**Comment by avakhrenev** [comment]
I'm not sure, but seems like scala changes the way how it loads jar files. Wierd things start to happen if jar file is modified in place. Looks like it somhow uses old `ZipEntry` pointers to access updated jar content. This seems relevant: [#3090](https://github.com/sbt/sbt/issues/3090#issuecomment-297174436)

**Comment by mpollmeier** [comment]
I used git bisect and traced it down to this commit: https://github.com/scala/scala/commit/6cb50acfb5ee4df342e83d8505116d4607

It introduces a `VirtualDirectoryFlatClassPath`, so we're getting closer. I filed a separate issue: https://github.com/scala/bug/issues/10541

**Comment by drxcc** [comment]
Using sbt-1.1.0-RC1 it seems to work now.  I tested it with scala-2.12.4  If it was a scala issues, maybe sbt server now reloads the entire jar file in memory.

**Comment by mpollmeier** [comment]
thanks @drxcc, I can confirm that it's fixed in 1.1.0-RC1 in combination with scala 2.12.4. I first tested with 2.12.3 and there it is/was still broken. anyway, from my perspective we can close this.
